projectq/backends/_sim/_classical_simulator.py

```diff
@@ -0,0 +1,131 @@
+"""
+A simulator that only permits classical operations, for faster/easier testing.
+"""
+
+from projectq.cengines import BasicEngine
+from projectq.ops import (XGate,
+                          BasicMathGate,
+                          Measure,
+                          FlushGate,
+                          Allocate,
+                          Deallocate)
+
+
+class ClassicalSimulator(BasicEngine):
+    """
+    A simple introspective simulator that only permits classical operations.
+
+    Allows allocation, deallocation, measuring (no-pop), flushing (no-op),
+    controls, NOTs, and any BasicMathGate. Supports reading/writing directly
+    from/to bits and registers of bits.
+    """
+    def __init__(self):
+        BasicEngine.__init__(self)
+        self._state = 0
+        self._bit_positions = {}
+
+    def read_bit(self, qubit):
+        """
+        Reads a bit.
+
+        Args:
+            qubit (projectq.types.Qubit): The bit to read.
+
+        Returns:
+            int: 0 if the target bit is off, 1 if it's on.
+        """
+        p = self._bit_positions[qubit.id]
+        return (self._state >> p) & 1
+
+    def write_bit(self, qubit, value):
+        """
+        Resets/sets a bit to the given value.
+
+        Args:
+            qubit (projectq.types.Qubit): The bit to write.
+            value (bool|int): Writes 1 if this value is truthy, else 0.
+        """
+        p = self._bit_positions[qubit.id]
+        if value:
+            self._state |= 1 << p
+        else:
+            self._state &= ~(1 << p)
+
+    def read_register(self, qureg):
+        """
+        Reads a group of bits as a little-endian integer.
+
+        Args:
+            qureg (projectq.types.Qureg):
+                The group of bits to read, in little-endian order.
+
+        Returns:
+            int: Little-endian register value.
+        """
+        return sum(self.read_bit(qureg[i]) << i
```

Comment:
I'd suggest using `result |= (value << i)` in a loop (it's almost 2x faster than summation)